cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+
+### Malicious Files
+
+The attacker delivers a file for malicious intent, such as:
+
+1. Exploit vulnerabilities in the file parser or processing module (_e.g._ [ImageTrick Exploit](https://imagetragick.com/), [XXE](https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Processing))
+2. Use the file for phishing (_e.g._ careers form)
+3. Send ZIP bombs, XML bombs (otherwise known as billion laughs attack), or simply huge files in a way to fill the server storage which hinders and damages the server's availability
+4. Overwrite an existing file on the system
+5. Client-side active content that could lead to all types of client-side attacks if the file is publicly retrievable
+6. Make the user upload a file without the user's consent, by crafting a Cross-Site Request Forgery attack (CSRF).
+
+### Public File Retrieval
+
+If the file uploaded is publicly retrievable, additional threats can be addressed:
+
+1. Public disclosure of other files
+2. Initiate a DoS attack by requesting lots of files. Requests are small, yet responses are much larger
+3. File content that could be deemed as illegal, offensive, or dangerous (_e.g._ personal data, copyrighted data, etc.) which will make you a host for such malicious files.
+
+## File Upload Protection
+
+There is no silver bullet in validating user content. Implementing a defense in depth approach is key to make the upload process harder and more locked down to the needs and requirements for the service. Implementing multiple techniques is key and recommended, as no one technique is enough to secure the service.
+
+### Extension Validation
+
+Ensure that the validation occurs after decoding the file name, and that a proper filter is set in place in order to avoid certain known bypasses, such as the following:
+
+- Double extensions, _e.g._ `.jpg.php`, where it circumvents easily the regex `\.jpg`
+- Null bytes, _e.g._ `.php%00.jpg`, where `.jpg` gets truncated and `.php` becomes the new extension
+- Generic bad regex that isn't properly tested and well reviewed. Refrain from building your own logic unless you have enough knowledge on this topic.
+
+Refer to the [Input Validation CS](Input_Validation_Cheat_Sheet.md) to properly parse and process the extension.
+
+#### Whitelist Extensions
+
+Ensure the usage of *business-critical* extensions only, without allowing any type of *non-required* extensions. If the system requires:
+
+- image upload, allow one type that is agreed upon to fit the business requirement;
+- cv upload, allow `docx` and `pdf` extensions.
+
+Based on the needs of the application, ensure the **least harmful** and the **lowest risk** file types to be used.
+
+#### Blacklist Extensions
+
+> Blacklisting extensions is a bad idea and is very dangerous. Don't do it unless you have no other choice!
```

Comment:
This syntax looks wrong to me - this should be used for quoting another document, and that's not what you're doing here. I'd just have this as a normal sentence (given there's only two sentences in this section).